### ISSUE TITLE
The random number generation and the decorator pattern

### ISSUE THREAD

**Issue by ajnebro** [jMetal/jMetal#114]
I have found the opportunity to apply the observer pattern in jMetal 5. To deal with random numbers we have the [PseudoRandomGenerator](https://github.com/jMetal/jMetal/blob/master/jmetal-core/src/main/java/org/uma/jmetal/util/pseudorandom/PseudoRandomGenerator.java) interface:

``` java
public interface PseudoRandomGenerator extends Serializable {
  public int nextInt(int lowerBound, int upperBound) ;
  public double nextDouble(double lowerBound, double upperBound) ;
  public double nextDouble() ;
  public void setSeed(long seed) ;
  public long getSeed() ;
  public String getName() ;
}
```

and a number of implementations of it (`JavaRandomGenerator`, `MersenneTwisterGenerator`, etc).

Now I'm dealing with an algorithm which requires two new methods in the random generator: `randNormal()` and `randSphere()`, so it seems that the observer pattern can be useful here, in this way:

``` java
package org.uma.jmetal.util.pseudorandom.impl;

import org.uma.jmetal.util.pseudorandom.PseudoRandomGenerator;

/**
 * Extended pseudo random number generator based on the decorator pattern.
 * Two new methods are added: randNormal() and randSphere()
 *
 * @author Antonio J. Nebro <antonio@lcc.uma.es>
 */
public class ExtendedPseudoRandomGenerator implements PseudoRandomGenerator {
  private final PseudoRandomGenerator randomGenerator ;

  public ExtendedPseudoRandomGenerator(PseudoRandomGenerator randomGenerator) {
    this.randomGenerator = randomGenerator ;
  }

  @Override
  public int nextInt(int lowerBound, int upperBound) {
    return randomGenerator.nextInt(lowerBound, upperBound) ;
  }

  @Override
  public double nextDouble(double lowerBound, double upperBound) {
    return randomGenerator.nextDouble(lowerBound, upperBound);
  }

  @Override
  public double nextDouble() {
    return randomGenerator.nextDouble();
  }

  @Override
  public void setSeed(long seed) {
    randomGenerator.setSeed(seed);
  }

  @Override
  public long getSeed() {
    return randomGenerator.getSeed() ;
  }

  @Override
  public String getName() {
    return randomGenerator.getName() + " (extended)";
  }

  /**
   * Use the polar form of the Box-Muller transformation to obtain
   * a pseudo random number from a Gaussian distribution
   * Code taken from Maurice Clerc's implementation
   *
   * @param mean
   * @param standardDeviation
   * @return A pseudo random number
   */
  public double randNormal(double mean, double standardDeviation) {
    double x1, x2, w, y1;

    do {
      x1 = 2.0 * randomGenerator.nextDouble() - 1.0;
      x2 = 2.0 * randomGenerator.nextDouble() - 1.0;
      w = x1 * x1 + x2 * x2;
    } while (w >= 1.0);

    w = Math.sqrt((-2.0 * Math.log(w)) / w);
    y1 = x1 * w;
    y1 = y1 * standardDeviation + mean;
    return y1;
  }

  /**
   * Get a random point from an hypersphere (center = 0, radius = 1)
   * Code taken from Maurice Clerc's implementation
   *
   * @param dimension
   * @return A pseudo random point
   */
  public double[] randSphere(int dimension) {
    int D = dimension;
    double[] x = new double[dimension];

    double length = 0;
    for (int i = 0; i < dimension; i++) {
      x[i] = 0.0;
    }

    // --------- Step 1. Direction

    for (int i = 0; i < D; i++) {
      x[i] = this.randNormal(0, 1);
      length += length + x[i] * x[i];
    }

    length = Math.sqrt(length);

    // --------- Step 2. Random radius

    double r = randomGenerator.nextDouble(0, 1);

    for (int i = 0; i < D; i++) {
      x[i] = r * x[i] / length;
    }

    return x;
  }

  /**
   * Ger a random point from an hypersphere
   * Code taken from Maurice Clerc's implementation
   *
   * @param center
   * @param radius
   * @return A pseudo random number
   */
  public double[] randSphere(int dimension, double center, double radius) {
    int d = dimension;
    double[] x = new double[dimension];

    double length = 0;
    for (int i = 0; i < dimension; i++) {
      x[i] = 0.0;
    }

    // --------- Step 1. Direction

    for (int i = 0; i < d; i++) {
      x[i] = randNormal(0, 1);
      length += length + x[i] * x[i];
    }

    length = Math.sqrt(length);

    // --------- Step 2. Random radius

    double r = randomGenerator.nextDouble(0, 1);

    for (int i = 0; i < d; i++) {
      x[i] = center + radius * r * x[i] / length;
    }

    return x;
  }
}
```

The point is that to ensure that there is only one instance of the random number generator we use the [JMetalRandom](https://github.com/jMetal/jMetal/blob/master/jmetal-core/src/main/java/org/uma/jmetal/util/pseudorandom/JMetalRandom.java) class, which applies the singleton pattern:

``` java
public class JMetalRandom implements Serializable {
  private static JMetalRandom instance ;
  private PseudoRandomGenerator randomGenerator ;

  private JMetalRandom() {
    randomGenerator = new JavaRandomGenerator() ;
  }

  public static JMetalRandom getInstance() {
    if (instance == null) {
      instance = new JMetalRandom() ;
    }
    return instance ;
  }

  public void setRandomGenerator(PseudoRandomGenerator randomGenerator) {
    this.randomGenerator = randomGenerator;
  }

  public int nextInt(int lowerBound, int upperBound) {
    return randomGenerator.nextInt(lowerBound, upperBound) ;
  }

  public double nextDouble() {
    return randomGenerator.nextDouble() ;
  }

  public double nextDouble(double lowerBound, double upperBound) {
    return randomGenerator.nextDouble(lowerBound, upperBound) ;
  }

  public void setSeed(long seed) {
    randomGenerator.setSeed(seed);
  }

  public long getSeed() {
    return randomGenerator.getSeed() ;
  }

  public String getGeneratorName() {
    return randomGenerator.getName() ;
  }
}
```

The problem arising now is how to combine this class with the one using the observer pattern.

**Comment by matthieu-vergne** [comment]
I don't see observers here (also called listeners). In the Javadoc of `ExtendedPseudoRandomGenerator` you mention the decorator pattern, which would make more sense.

What I see here is that `ExtendedPseudoRandomGenerator` acts as a decorator, so classes which already have a random generator but need these additional methods can instantiate the decorator by giving it the random generator they already have.

What I understand by reading your post is that you see how to apply the decorator pattern to this case, but you don't see how to use it. It should be the reverse : you face a situation, and there is a pattern that is usually used in this kind of situation, justifying that you use it too.

What I see with the singleton is that you expect to use the same random generator everywhere (the one stored in `JMetalRandom`), but which can change over time (through `JMetalRandom.setRandomGenerator(...)`). In such a case, you always need to call `JMetalRandom.getInstance()` and use its own methods, so making a decorator -which stores the random generator to use- does not seem compliant with the initial assumption of using a singleton. In particular, the decorator pattern implies that you use the decorator, not the singleton (which is contained in the decorator). It goes against the idea of using a singleton: you might implement several times the decorator, each using the initial singleton but none of them *being* a singleton anymore.

Do you really need to have a singleton? What motivated that?

**Comment by matthieu-vergne** [comment]
So far, it seems more interesting to me to create a class which has only the three additional methods, each of them calling the single instance from `JMetalRandom`, without containing any generator nor implementing `PseudoRandomGenerator`. When you just need a normal generator, use the single instance from `JMetalRandom`, and when you need these additional methods simply instantiate it. You might even use a singleton pattern on it.

**Comment by ajnebro** [comment]
> I don't see observers here

I wrote observer instead of decorator :-|

> Do you really need to have a singleton? What motivated that?

Yes. The random generator is widely used in metaheuristics, and there must be only one instance of it to allow to fix the seed and this way to be able to reproduce the executions.

**Comment by matthieu-vergne** [comment]
I would actually say the opposite: if you want the generator to follow the same path, you should provide it the same seed... but you should also ensure that not external entity uses it at the same time, otherwise it may consume random numbers that you will miss in your re-execution. If an algorithm needs a random generator, it instantiates one for itself and uses this one. If you run several algorithms in parallel threads, all of them using the same random generator, be sure to never be able to re-execute properly your sequence.

**Comment by ajnebro** [comment]
> I would actually say the opposite: if you want the generator to follow the same path, you should provide it the same seed.

I use concrete seeds in two cases: for debugging and to analyze some executions providing interesting results. In fact, a colleague told me that he always printed the seed in the screen when running any algorithm; I'm following his advice since then.

> If you run several algorithms in parallel threads, all of them using the same random generator, be sure to never be able to re-execute properly your sequence.

When running multithreaded or distributed algorithms there is no way to repeat the behaviour of particular executions. This is one of the reasons why parallel programming is "so funny".

**Comment by matthieu-vergne** [comment]
Yes it is. You have to control the process enough to make it deterministic. At least on the properties you are analysing. This is for instance a problem for debugging online games like MMORPG: the clients and the server should be synchronized on some aspects, and pausing one for debugging implies to pause the other to avoid timeout disconnections or crashes. I don't say it is easy, but yes it is feasible and already done in some domains. It does not mean it should necessarily be fully synchronized (which is really hard and implies coding tricks but is rarely useful) but synchronizing on relevant steps can be done (and sometimes must be done anyway). And there is architectures and features for that.

When you use a singleton, this is because you have a global uniqueness requirement (everyone must use the very same instance everywhere), or because of performance optimization purpose. I don't see the uniqueness requirement for a random generator (at the opposite of having a unique thread manager to ensure the use of an homogeneous policy, or a unique GUI because there is actually only one GUI), and the performance issue does not appear to me to be significant (I expect to see 1 generator per algorithm instance, so not much compared to the number of solutions). I understand that a random generator might be unique within a given scope (e.g. an algorithm should use the same for all of its tasks), but in this case it means that you should instantiate the generator at this scope level (e.g. at the same level where you instantiate the algorithm). Singleton are for the most global scope, where no code applies yet (`static`). Using it for the random generator seems to me going too far: when you run your program, you instantiate a random generator that you give to your algorithm, or that you store in some setting structures that you give to your algorithm (or other objects). If it is all what you need, then the singleton is nothing more than a trick to minimize the code changes, but it is not the best solution and it might shows limitations later, when you want to use a different generator instance (possibly with the same seed to keep it deterministic), for instance in order to avoid locks in parallel threads by providing each of them its own instance.

Well, this is not the main topic of the issue, so let's close that point. You might think about it if you want and open another issue if relevant, it does not block the decorator issue anyway so let's come back to that.

The decorator might help you here if you provide a basic generator at a high level (singleton or not) that you would like to "extend" in order to have more specific random features within the algorithm. The first idea is just to create methods in your algorithm that will use the generator, but if you want to reuse these methods elsewhere then they should not be defined in the algorithm (or they may be static, but you still introduce a dependence between different algorithms, so not recommended). To solve this problem, you normally extend the random generator, but the problem is that the one who decide which generator to provide (high level, e.g. generic algorithm runner) does not know which specific class to use (low level, algorithm-specific info), so you would need low level info to be shared at a higher level, which implies more complexity, mess, etc. To avoid that, you need to define these additional methods in an independent class that the low level can use upon its own decision, and such that you can reuse the high level generator (because the high level still ask you to use this one). It does not necessarily have to be a decorator (it does not have to extend the same interface in order to provide the same methods), but if you need also the basic methods then you will need to use 2 instances (the high level generator + the new instance which builds on it). Making it a decorator allows you to have both the high level and low level methods in the new instance. And because it extends the same interface, for methods which still ask for the high level generator you can provide your algorithm-specific decorator. So you can encapsulate the high level generator into your decorator and use only your decorator everywhere in the algorithm.

Consequently, the decorator acts at the low level (algorithm-specific), and you might encapsulate in further decorators when going deeper in the specificity. These decorators applies only at these levels and do not impact higher levels (other algorithms must use their own decorators, even if they are the same). If your idea is to add features somewhere that can be used elsewhere (so make them available not only for lower levels) then you need to use a different architecture.
